feat: otter sdk training - model extension (#2411)
AmadeusITGroup/otter: apps/showcase/src/assets/trainings/sdk/steps/model-extension/instructions.md

There are certain cases in which you want to be able to extend your SDK models, and therefore ensure that revivers are generated.

Let's continue with the use case of the previous exercise.\
In order to keep track of the user's current booking, it would be useful to generate an ID.
To do this, we can create a new model which extends the previously generated `Flight` type.

> [!NOTE]

> > For this exercise, we provide an SDK that has been generated with model extension.

Author: ```suggestion
> This exercise requires an SDK that has been generated with model extension, which we have provided.
```

> 
> To verify this, check out the configuration used to generate the API (either the command line or the file `openapitools.json`).\
> Ensure that the global property `allowModelExtension` has been set to `true`.\
> This will guarantee that the revivers of the base models are generated, which is essential for the following exercise.

### Exercise

Extensions of base models (located in the `base` folder) are created in the `core` folder.

#### Creating the extended model
First, let's create the type `FlightCoreIfy` in `libs/sdk/src/models/core/flight.ts`.

> This type should extend the type `Flight`, imported from the `base` folder and add a new field `id: string`.

Author: just a suggestion so the user writes the field themselves
```suggestion
This type should extend the type `Flight`, imported from the `base` folder and add a new field `id` of type `string`.
```

Author: should we add an example or a template here in the instructions?

Reviewer: I want to avoid the exercise being just some copy-paste and replace.
Unless you think this step may be too difficult for some people, I think it's best to let them figure it out.

> [!WARNING]
> The naming convention requires the core model to contain the suffix `CoreIfy`.\
> You can find more information on core models in the
> <a href="https://github.com/AmadeusITGroup/otter/blob/main/docs/api-sdk/SDK_MODELS_HIERARCHY.md" target="_blank">SDK models hierarchy documentation</a>.

#### Creating the extended reviver
Then, you can create the reviver for this new core model in `libs/sdk/src/models/core/flight.reviver.ts`.\

> This extended reviver will call the reviver of the base `Flight` model and add the `id` to the returned object.\

Author: same thing as extended model, should we add an example or a template?

Reviewer: same answer, you may provide some hints but not a full solution to copy paste

> The base reviver should exist since we have ensured this during the prerequisites of the exercise.

Author: maybe move the `[!NOTE]` to the `### Exercise` section and add **Prerequisite** at the beginning of the note to be more clear?

Reviewer: I've created a dedicated section at the beginning of the exercise to explain the base model and reviver, so I've removed this sentence

#### Updating the exports
Once the core model and its reviver are created, we can go back to the base model to update the exported models and revivers.\

> Update the file `libs/sdk/src/models/base/flight/index.ts` to export your extended model and reviver instead of the original.

Author: the new `index.ts` file should export the extended model, reviver, and base model no?

Reviewer: Same as `IgnoreEnum`, exporting the base model is not strictly necessary to perform this exercise, so I prefer not to mention it.

In the folder `libs/sdk/src/models/core/flight`, you will see the same 3 files mentioned before.

First, let's create the type `FlightCoreIfy` in `libs/sdk/src/models/core/flight.ts`.
This type should extend the type `Flight`, imported from the `base` folder and add a new field `id` of type `string`.

> [!WARNING]
> The naming convention requires the core model to contain the suffix `CoreIfy`.\
> You can find more information on core models in the
> <a href="https://github.com/AmadeusITGroup/otter/blob/main/docs/api-sdk/SDK_MODELS_HIERARCHY.md" target="_blank">SDK models hierarchy documentation</a>.

#### Creating the extended reviver
Now that you have your extended model, let's create the associated reviver in `libs/sdk/src/models/core/flight.reviver.ts`.\
This extended reviver will call the reviver of the base `Flight` model and add the `id` to the returned object.

#### Updating the exports
Once the core model and its reviver are created, we can go back to the base model to update the exported models and revivers.\
Update the file `libs/sdk/src/models/base/flight/index.ts` to export your extended model and reviver instead of the original.

#### Seeing the result

> You should now have your extension working!\

Author: ```suggestion
Your extension should now be working!\
```

Check out the preview to see if the `id` has been added to the model.

> 

Reviewer 2: Should we mention the fact that we need to not regenerate the `libs/sdk/src/models/base/flight/index.ts`, meaning adding the path in the `.openapi-generator-ignore` file ?

This type should extend the type `Flight`, imported from the `base` folder and add a new field `id` of type `string`.

> [!WARNING]
> The naming convention requires the core model to contain the suffix `CoreIfy`.\
> You can find more information on core models in the
> <a href="https://github.com/AmadeusITGroup/otter/blob/main/docs/api-sdk/SDK_MODELS_HIERARCHY.md" target="_blank">SDK models hierarchy documentation</a>.

#### Creating the extended reviver
Now that you have your extended model, let's create the associated reviver in `libs/sdk/src/models/core/flight.reviver.ts`.\
This extended reviver will call the reviver of the base `Flight` model and add the `id` to the returned object.

#### Updating the exports
Once the core model and its reviver are created, we can go back to the base model to update the exported models and revivers.\
Update the file `libs/sdk/src/models/base/flight/index.ts` to export your extended model and reviver instead of the original.

#### Seeing the result
Your extension should now be working!\
Check out the preview to see if the `id` has been added to the model.

#### Persistence of the change

> You may have realised that we have modified a portion of code that was originally generated.

Author: ```suggestion
You may have realized that we have modified a portion of code that was originally generated.
```

We don't want to lose the change that we made on `libs/sdk/src/models/base/flight/index.ts` next time it's regenerated.

> To avoid that, we can add `src/models/base/flight/index.ts` to `libs/sdk/.openapi-generator-ignore` file.

Author: ```suggestion
To avoid that, we can add `src/models/base/flight/index.ts` to the `libs/sdk/.openapi-generator-ignore` file.
```

> We cannot regenerate the SDK in the code editor due to a Java dependency but fell free to try it in a local project.

Author: ```suggestion
We cannot regenerate the SDK in the code editor due to a Java dependency but fele free to try it in a local project.
```
